gpu_bdb/bdb_tools/q03_utils.py

```diff
@@ -111,19 +134,29 @@ def apply_find_items_viewed(df, item_mappings):
     size = len(sample)
 
     # we know this can be int32, since it's going to contain item_sks
-    out_arr = cuda.device_array(size * N, dtype=df["wcs_item_sk"].dtype)
-
-    find_items_viewed_before_purchase_kernel.forall(size)(
-        sample["relevant_idx_pos"],
-        df["wcs_user_sk"],
-        df["tstamp"],
-        df["wcs_item_sk"],
-        out_arr,
-        N,
-    )
-
-    result = cudf.DataFrame({"prior_item_viewed": out_arr})
-
+    if isinstance(df, cudf.DataFrame):
+        out_arr = cuda.device_array(size * N, dtype=df["wcs_item_sk"].dtype)
+        find_items_viewed_before_purchase_kernel_gpu.forall(size)(
+            sample["relevant_idx_pos"],
+            df["wcs_user_sk"],
+            df["tstamp"],
+            df["wcs_item_sk"],
+            out_arr,
+            N,
+        ) 
+        result = cudf.DataFrame({"prior_item_viewed": out_arr})
+    else: 
+        out_arr = np.zeros(size * N, dtype=df["wcs_item_sk"].dtype, like=df["wcs_item_sk"].values)
```

Comment:
This can be common across both GPU because by using the like variable it will automatically dispatch to cupy array/numpy array as relevant.